File: matlab/+perceptron/minover.m
%% Minover
function [weights, bias, should_stop] = minover(samples, labels, weights, bias, threshold)
    import perceptron.hebbian_learning_step;
    import perceptron.lowest_stability;
    should_stop = false;

    % do a hebbian learning step using the sample with lowest stability
    [stability, idx] = lowest_stability(weights, samples, labels);
    [new_weights, bias] = hebbian_learning_step( ...
        weights, ...
        bias, ...
        samples(idx), ...
        labels(idx), ...
        size(samples, 2) ...
    );

    previous_stability = stability + 1;

    if abs(1 - new_weights'* weights / norm(weights) / norm(new_weights)) < 1E-5 
        should_stop = true;
    end
    weights = new_weights;

    % if (norm(stability - previous_stability) < threshold)
    %     should_stop = true;
    % end
end

File: neural-cpp/src/Perceptrons/Minover.cpp
#include "Minover.h"
#include <cmath>

Minover::Minover(Dichotomy &dichotomy)
  :
  Perceptron<Minover>(dichotomy)
{}

void Minover::train(Dichotomy &dichotomy)
{
    auto idx = find_minimal_stability_index();
    rememberStability(stability(idx));

    auto sample = dichotomy.samples[idx];
    auto label = dichotomy.labels[idx];

    hebbian_learning_step(sample, label);
}

void Minover::rememberStability(double stability)
{
    d_stabilities.push(stability);
    // Only remember the last P stabilities
    if (d_stabilities.size() > d_weights.size())
        d_stabilities.pop();
}

bool Minover::converged() const
{
    auto current_stability = d_stabilities.back();
    auto previous_stability = d_stabilities.front();

    return std::abs(current_stability - previous_stability) < 1e-5;
}
